### PR TITLE
Add bootstrap.php checking environment (MB_OVERLOAD_STRING).

### LINKED ISSUE #762: Handling mbstring.func_overload
PHP has this really stupid feature called `mbstring.func_overload` which affects the behavior of `strlen()` and `substr()`.

Normally, these functions operate on a per-byte basis, but if you enable `mbstring.func_overload` it adopts multibyte character support. The consequences can be insane (e.g. string lengths being calculated based off a UTF-8 interpretation of a string, which can degrade security).

defuse/php-encryption solves this by using `mb_strlen($string, '8bit')` and `mb_substr($string, $start, $length, '8bit');` if these functions exist, and falls back to the defaults if they do not.

https://github.com/defuse/php-encryption/blob/d2c33279e31d5f3d4b74a82139b6ce56d551965b/src/Crypto.php#L654-L709

It might be worthwhile to adopt a similar approach in phpseclib.

### PR DESCRIPTION
**Pull request by bantu**
Fix #762

### CODE REVIEW

**Comment by terrafrost** [comment]
I guess the problem I have with this approach is that phpseclib doesn't have to be installed via composer.

**Comment by bantu** [comment]
Let's just make everything else unsupported.

**Comment by terrafrost** [comment]
I don't know that I completely support that.

Last month 2,447 came from sourceforge.net per http://sourceforge.net/projects/phpseclib/files/stats/timeline?dates=2011-07-29+to+2015-08-04

3.8 million came from packagist.org but that's all time and a lot of that is likely people who didn't intentionally download it.

A lot of PHP libraries can be downloaded and utilized without Composer. They may not be very good libraries but I think none-the-less a good number of PHP devs (altho perhaps not very good ones) do expect to be able to use stuff without Composer.

That said, I suppose it is easy enough to get this working without composer (for the 2.0.0 branch and later). ie.

```php
<?php
include('autoload.php');
include('bootstrap.php');

use phpseclib\Net\SSH2;

$ssh = new SSH2();
```

And I guess people can still download the 1.0 branch. Altho after X509 and SSH2 have had an API revamp I plan on redoing the website to promote whatever the latest version is (be it 2.1, 2.2 or whatever) I do figure I'll stop mentioning the 1.0 branch on the website.

Long story short, I guess I'm okay making other things unsupported.

Altho that said, wouldn't making everything else unsupported be a BC breaking change? ie. maybe this should go in master and not 2.0? And fwiw this proposed fix wouldn't fix it in the 1.0 branch, altho tbh, I don't think that's a huge issue anyway. Fact is SSH2.php (and probably the symmetric ciphers too) wouldn't even work with MB_OVERLOAD_STRING anyway.

**Comment by bantu** [comment]
phpseclib does not support MB_OVERLOAD_STRING in any version, so this is clearly not a BC break.

First of all, we need to document this incompatibility, which I'd say is even sufficient to deal with this problem.

Then, loading the bootstrap file via composer is an optional precaution to make it easier for most of our users. This optional precaution could potentially also go into the 1.0 branch.

File: phpseclib/bootstrap.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Bootstrapping File for phpseclib
+ *
+ * @license http://www.opensource.org/licenses/mit-license.html MIT License
+ */
+
+if (extension_loaded('mbstring')) {
+    // 2 - MB_OVERLOAD_STRING
+    if (ini_get('mbstring.func_overload') & 2) {
+        throw new \UnexpectedValueException(
```
Comment (GrahamCampbell) [inline phpseclib/bootstrap.php]: don't need the slasg

File: phpseclib/bootstrap.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<?php
+/**
+ * Bootstrapping File for phpseclib
```
Comment (GrahamCampbell) [inline phpseclib/bootstrap.php]: missing fullstop

**Comment by terrafrost** [comment]
> phpseclib does not support MB_OVERLOAD_STRING in any version, so this is clearly not a BC break.

No, but not supporting anything other than composer installs - as you proposed - is essentially a BC break.

I don't have a problem with it being a little bonus compatibility check for people using composer but to say that the only way phpseclib will be supported hence-forth is by composer installs...  that is essentially a BC break.

**Comment by bantu** [comment]
@terrafrost I see.

File: phpseclib/bootstrap.php (hunk 1 above)
Comment (paragonie-scott) [inline phpseclib/bootstrap.php]: Since this is not in a namespace, I agree, but there's no harm in being explicit.

**Comment by Joey3000** [comment]
> A lot of PHP libraries can be downloaded and utilized without Composer. They may not be very good libraries but I think none-the-less a good number of PHP devs (altho perhaps not very good ones) do expect to be able to use stuff without Composer.

:+1: Phpseclib can be used e.g. in plugins for CMS/Apps potentially used on shared hosts. E.g., [for Joomla](http://extensions.joomla.org/extension/encrypt-configuration) and [for Piwik](https://github.com/Joey3000/piwik-LoginEncrypted).

File: phpseclib/bootstrap.php (hunk 1 above)
Comment (GrahamCampbell) [inline phpseclib/bootstrap.php]: but that's not being explicit. class names are defined without a slash
